GVFS/GVFS.Common/GVFSEnlistmentHealthCalculator.cs

```diff
@@ -0,0 +1,240 @@
+                }
+
+                // Only categorize if descendent of the parentDirectory
+                if (formattedPath.StartsWith(parentDirectory))
+                {
+                    count++;
+
+                    if (isFile)
+                    {
+                        // Find out the top level directory of the files path, which will be one level under the parent directory
+                        // If the file is in the parent directory, topDir is set to '/' (just temporary, formatting will eventually just show the path to the parent directory)
+                        string topDir = this.ParseTopDirectory(this.TrimDirectoryFromPath(formattedPath, parentDirectory));
+                        this.IncreaseDictionaryCounterByKey(directoryTracking, topDir);
+                    }
+                    else
+                    {
+                        // If the path is to the parentDirectory, ignore it to avoid adding string.Empty to the data structures
+                        if (!parentDirectory.Equals(formattedPath))
+                        {
+                            // Trim the path to parent directory from the path to this directory
+                            string topDir = this.TrimDirectoryFromPath(formattedPath, parentDirectory);
+
+                            // If this directory isn't already one level under the parent...
+                            if (topDir.IndexOf(GVFSConstants.GitPathSeparator) != -1)
+                            {
+                                // ... Get the one that is
+                                topDir = this.ParseTopDirectory(topDir);
+                            }
+
+                            this.IncreaseDictionaryCounterByKey(directoryTracking, topDir);
+                        }
+                    }
+                }
+            }
+
+            return count;
+        }
+
+        /// <summary>
+        /// Trim the relative path to a directory from the front of a specified path which is its child
+        /// </summary>
+        /// <remarks>Precondition: Only even run in the context of TrimDirectoryFromPath(child, parentDirectory)</remarks>
+        /// <param name="path">The path being trimmed</param>
+        /// <param name="directoryTarget">The directory target whose path to trim from the path</param>
+        /// <returns>The newly formatted path with the directory trimmed</returns>
+        private string TrimDirectoryFromPath(string path, string directoryTarget)
+        {
+            return path.Substring(directoryTarget.Length);
+        }
+
+        private void IncreaseDictionaryCounterByKey(Dictionary<string, int> countingDictionary, string key)
+        {
+            if (!countingDictionary.TryGetValue(key, out int count))
+            {
+                count = 0;
+            }
+
+            countingDictionary[key] = ++count;
+        }
+
+        public class GVFSEnlistmentHealthData
+        {
+            public GVFSEnlistmentHealthData(

```

Comment:
There is a lot of data that we want to have for each "directory" - whether the top level directory or its children. Should we have a data structure to represent hold this common data?

```
public DirectoryEnlistmentHealthData
{
    public TrackedFileCount { get; }
    public PlaceholderCount { get; }
    public ModifiedPathCount { get; }
    public HydrationLevel { get; }
}
```

The top level result could then be something like:

```
public class GVFSEnlistmentHealthData
{
    string Directory { get; }
    DirectoryEnlistmentHealthData HealthData { get; }
    Public Dictionary<string, DirectoryEnlistmentHealthData> ChildDirectoryHealthData { get; }
}
```

Reply:
My intention was to only have a lot of data for the highest level directory, either `src` or whatever directory was pointed at with the `-d` flag. The directories below that don't have the detailed information such individual counts, just the `HydrationLevel` metric.